### PR TITLE
Create AST node to represent AT TIME ZONE expressions

### PR DESCRIPTION
**Pull request by geraint0923**
Address #4289

### CODE REVIEW

**Comment by geraint0923** [comment]
Implemented AT TIME ZONE as an Expression and applied the changes we discussed.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    public AtTimeZone(NodeLocation location, Expression valueExpression, Expression timeZoneSpecifier)
+    {
+        this(Optional.of(location), valueExpression, timeZoneSpecifier);
+    }
+
+    private AtTimeZone(Optional<NodeLocation> location, Expression valueExpression, Expression timeZoneSpecifier)
+    {
+        super(location);
+        this.valueExpression = valueExpression;
```
Comment (erichwang) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: requireNonNull(valueExpression, "valueExpression is null);

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        this.timeZoneSpecifier = timeZoneSpecifier;
```
Comment (erichwang) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: requireNonNull

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java (hunk 1 above)
Comment (erichwang) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: in general, we always try to check for null values on constructor setting of fields

File: presto-parser/src/main/java/com/facebook/presto/sql/ExpressionFormatter.java [hunk 3]
```diff
@@ -121,6 +122,16 @@ protected String visitExpression(Expression node, Boolean unmangleNames)
         }
 
         @Override
+        protected String visitAtTimeZone(AtTimeZone node, Boolean context)
+        {
+            StringBuilder builder = new StringBuilder();
```
Comment (erichwang) [inline presto-parser/src/main/java/com/facebook/presto/sql/ExpressionFormatter.java]: you can chain this whole thing into one fluent statement:

```java
return new StringBuilder()
    .append(...)
    .append(...)
    .toString()
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -742,6 +743,25 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<A
...
         @Override
+        protected Type visitAtTimeZone(AtTimeZone node, StackableAstVisitorContext<AnalysisContext> context)
+        {
+            Type valueType = process(node.getValueExpression(), context);
+            process(node.getTimeZoneSpecifier(), context);
+            if (!valueType.equals(TIME_WITH_TIME_ZONE) && !valueType.equals(TIMESTAMP_WITH_TIME_ZONE)) {
+                throw new SemanticException(TYPE_MISMATCH, node.getValueExpression(), "Type of value must be TIME_WITH_TIME_ZONE or TIMESTAMP_WITH_TIME_ZONE (actual %s)", valueType);
+            }
+            Expression timeZoneSpecifier = node.getTimeZoneSpecifier();
+            if (!(timeZoneSpecifier instanceof IntervalLiteral) && !(timeZoneSpecifier instanceof StringLiteral)) {
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: not to clear on the spec here, but should we be checking for the Type here instance of instanceof? Or does it strictly have to be one of these two literals?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -742,6 +743,25 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<A
...
+                throw new SemanticException(TYPE_MISMATCH, node.getValueExpression(), "Timezone specifier must be INTERVAL or STRING (actual %s)", timeZoneSpecifier.getClass());
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: STRING => VARCHAR

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 6]
```diff
@@ -742,6 +743,25 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<A
...
+            }
+            Type resultType = valueType.equals(TIME_WITH_TIME_ZONE) ? TIME_WITH_TIME_ZONE : TIMESTAMP_WITH_TIME_ZONE;
+            expressionTypes.put(node, resultType);
+            // TODO: is it necessary to resolve "at_time" here and add to resolvedFunctions
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: no we replace with the function later

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    public AtTimeZone(NodeLocation location, Expression valueExpression, Expression timeZoneSpecifier)
+    {
+        this(Optional.of(location), valueExpression, timeZoneSpecifier);
+    }
+
+    private AtTimeZone(Optional<NodeLocation> location, Expression valueExpression, Expression timeZoneSpecifier)
+    {
+        super(location);
+        this.valueExpression = requireNonNull(valueExpression, "AtTimeZone valueExpression is null");
```
Comment (erichwang) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: Traditionally, we just put "valueExpression is null". This is autogenerated for the most part.

**Comment by geraint0923** [comment]
Addressed comments, translate AtTimeZone to FunctionCall after planing as an optimizer

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableList;
+
+public class DesugaringRewriter
+        extends ExpressionRewriter<Void>
+{
+    public DesugaringRewriter()
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java]: don't need the constructor

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+import java.util.Map;
+
+public class DesugaringOptimizer
+        extends PlanOptimizer
+{
+    public DesugaringOptimizer()
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java]: constructor not needed

**Comment by erichwang** [comment]
last few comments, otherwise looks good

**Comment by geraint0923** [comment]
Thanks a lot!
Just remove the unnecessary constructors to address the comments.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 10]
```diff
@@ -742,6 +743,20 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<A
...
 
         @Override
+        protected Type visitAtTimeZone(AtTimeZone node, StackableAstVisitorContext<AnalysisContext> context)
+        {
+            Type valueType = process(node.getValueExpression(), context);
+            process(node.getTimeZoneSpecifier(), context);
+            if (!valueType.equals(TIME_WITH_TIME_ZONE) && !valueType.equals(TIMESTAMP_WITH_TIME_ZONE)) {
+                throw new SemanticException(TYPE_MISMATCH, node.getValueExpression(), "Type of value must be TIME_WITH_TIME_ZONE or TIMESTAMP_WITH_TIME_ZONE (actual %s)", valueType);
+            }
+            Type resultType = valueType.equals(TIME_WITH_TIME_ZONE) ? TIME_WITH_TIME_ZONE : TIMESTAMP_WITH_TIME_ZONE;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Isn't this equivalent to `Type resultType = valueType`?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class AtTimeZone
+        extends Expression
+{
+    private final Expression valueExpression;
+    private final Expression timeZoneSpecifier;
+
+    public static final QualifiedName FUNCTION_NAME = QualifiedName.of("at_timezone");
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: This shouldn't be in the AST. It's an internal detail of the way the planner implements this operation.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    private AtTimeZone(Optional<NodeLocation> location, Expression valueExpression, Expression timeZoneSpecifier)
+    {
+        super(location);
+        checkArgument(timeZoneSpecifier instanceof IntervalLiteral || timeZoneSpecifier instanceof StringLiteral, "timeZoneSpecifier must be IntervalLiteral or StringLiteral");
+        this.valueExpression = requireNonNull(valueExpression, "valueExpression is null");
+        this.timeZoneSpecifier = requireNonNull(timeZoneSpecifier, "timeZoneSpecifier is null");
+    }
+
+    public Expression getValueExpression()
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: Call this just `getValue`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        return valueExpression;
+    }
+
+    public Expression getTimeZoneSpecifier()
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/AtTimeZone.java]: Call this just `getTimeZone`

**Comment by geraint0923** [comment]
Addressed the comments.

**Comment by erichwang** [comment]
Merged!

**Pull request closed without merging**